File: react-native-app/src/services/sessionBuilder.ts
import {
  EXERCISES,
  LEVELS,
  REST_LABEL_KEY,
  getCycleSeconds,
  getDurationBounds,
  getSteps,
  isFreeExercise,
} from '../constants/catalogues';

export interface PlaylistStep {
  phase: 'contract' | 'relax';
  /** i18n key for the cue inside the circle. Resolved by the screen, not here. */
  labelKey: string;
  seconds: number;
  from: number;
  to: number;
  /** Exercise this step belongs to, or 'rest'. Doubles as the identity used to
   *  group consecutive steps into one carousel block. */
  slug: string;
}

export interface Playlist {
  steps: PlaylistStep[];
  /** Slugs, in the order they are trained. Names are looked up for display. */
  exercises: string[];
  totalSeconds: number;
}

// Shuffles an array helper
const shuffle = <T>(array: T[]): T[] => {
  const arr = [...array];
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
};

// Calculate duration of an exercise for a specific level number
export const getDurationForLevel = (slug: string, levelNumber: number): number => {
  const cycle = getCycleSeconds(slug);
  if (cycle <= 0) return 30.0;

  const [min, max] = getDurationBounds(slug);
  const topNumber = 5; // Levels range from 1 to 5
  
  const pct = topNumber > 1 ? Math.max(0.0, Math.min(1.0, (levelNumber - 1) / (topNumber - 1))) : 0.0;
  const target = min + (max - min) * pct;

  let cycles = Math.max(1, Math.round(target / cycle));
  // Whole cycles, but never blow past the range maximum on rounding
  while (cycles > 1 && cycles * cycle > max + 0.01) {
    cycles--;
  }
  // ...and symmetrically, never fall short of the minimum.
  //
  // Only the maximum was guarded, so the duration range was enforced at one
  // end and not the other. Rounding to a whole number of cycles can land under
  // `min` whenever the cycle length does not divide it: front-clamp has a 3.3s
  // cycle and a 20s floor, so level 1 rounded to 6 cycles - 19.8s, a set
  // shorter than the catalogue says the exercise is. The extra cycle is only
  // taken when it still fits under `max`, so this can never fight the loop
  // above.
  while (cycles * cycle < min - 0.01 && (cycles + 1) * cycle <= max + 0.01) {
    cycles++;
  }

  return Math.round(cycles * cycle * 10) / 10;
};

// Generate daily workout session playlist
export const buildDailySession = (
  completedDays: number,
  levelNumber: number,
  opts?: { freeOnly?: boolean },
): Playlist => {
  const level = LEVELS[levelNumber] || LEVELS[1];
  const total = level.total_session_seconds;
  const rest = level.rest_seconds;

  // Filter to only unlocked exercises.
  //
  // A free account is bounded twice over, and both bounds matter. Its day
  // count stops at FREE_DAY_CAP so the day-based unlock can never reach the
  // fourth exercise, and this filter is the belt to that braces: even if a
  // day count arrived from somewhere else - a sync from an account that was
  // subscribed yesterday, a clock change - the session still only ever
  // contains what the tier actually includes.
  const unlocked = Object.values(EXERCISES).filter(
    (ex) =>
      completedDays >= ex.unlock_after_days
      && (!opts?.freeOnly || isFreeExercise(ex.slug)),
  );

  if (unlocked.length === 0) {
    return { steps: [], exercises: [], totalSeconds: 0 };
  }

  // Shuffle unlocked exercises
  let pool = shuffle(unlocked);
  const sequence: any[] = [];
  const durations: number[] = [];
  let acc = 0;
  let idx = 0;

  while (sequence.length < 30) {
    let candidate = pool[idx % pool.length];
    
    // Avoid two consecutive exercises of the same type if pool has more than 1
    if (sequence.length > 0 && sequence[sequence.length - 1].slug === candidate.slug && pool.length > 1) {
      idx++;
      candidate = pool[idx % pool.length];
    }

    const duration = getDurationForLevel(candidate.slug, levelNumber);
    const addition = (sequence.length === 0 ? 0.0 : rest) + duration;

    // Check if adding exceeds target total
    if (sequence.length > 0 && acc + addition > total) {
      const over = (acc + addition) - total;
      const under = total - acc;
      if (over >= under) {
        break; // Stopping here gets us closer to target
      }
    }

    durations.push(duration);
    sequence.push(candidate);
    acc += addition;
    idx++;

    if (acc >= total) {
      break;
    }
  }

  // Construct the final playlist steps
  const steps: PlaylistStep[] = [];
  const exercises: string[] = [];
  let finalAcc = 0;

  sequence.forEach((exercise, i) => {
    // Insert rest between exercises
    if (i > 0 && rest > 0) {
      steps.push({
        phase: 'relax',
        labelKey: REST_LABEL_KEY,
        seconds: rest,
        from: 0,
        to: 0,
        slug: 'rest',
      });
      finalAcc += rest;
    }

    // Insert movement steps
    const exerciseDuration = durations[i];
    const exerciseSteps = getSteps(exercise.slug, exerciseDuration);
    
    exerciseSteps.forEach((s) => {
      steps.push({
        phase: s.phase,
        labelKey: s.labelKey,
        seconds: s.seconds,
        from: s.from,
        to: s.to,
        slug: exercise.slug,
      });
      finalAcc += s.seconds;
    });

    if (!exercises.includes(exercise.slug)) {
      exercises.push(exercise.slug);
    }
  });

  return {
    steps,
    exercises,
    totalSeconds: Math.round(finalAcc * 10) / 10,
  };
};

// Generate single trial exercise workout playlist (for practicing a specific exercise)
export const buildSingleSession = (slug: string, levelNumber: number): Playlist => {
  const exercise = EXERCISES[slug];
  if (!exercise) {
    return { steps: [], exercises: [], totalSeconds: 0 };
  }

  const duration = getDurationForLevel(slug, levelNumber);
  const exerciseSteps = getSteps(slug, duration);

  const steps: PlaylistStep[] = exerciseSteps.map((s) => ({
    phase: s.phase,
    labelKey: s.labelKey,
    seconds: s.seconds,
    from: s.from,
    to: s.to,
    slug: exercise.slug,
  }));

  const total = steps.reduce((sum, s) => sum + s.seconds, 0);

  return {
    steps,
    exercises: [exercise.slug],
    totalSeconds: Math.round(total * 10) / 10,
  };
};

File: wear-app/app/src/main/java/com/kegelee/app/wear/SessionBuilder.kt
package com.kegelee.app.wear

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * One step of a session: hold or release, for this long, at this circle size.
 *
 * `from` and `to` are the circle's scale at the start and end of the step, 0..1.
 * They are what makes the training circle breathe rather than merely count, and
 * they are the reason this app draws the same thing the phone does instead of a
 * generic countdown.
 */
data class PlayStep(
    val phase: String,
    val label: String,
    val seconds: Double,
    val from: Double,
    val to: Double,
    val slug: String,
) {
    val isContract: Boolean get() = phase == "contract"
    val isRest: Boolean get() = slug == "rest"
}

data class Playlist(
    val steps: List<PlayStep>,
    val exercises: List<String>,
    val totalSeconds: Double,
)

/**
 * A direct port of the phone app's `services/sessionBuilder.ts`.
 *
 * Deliberately line-for-line rather than reinterpreted, down to the rounding
 * and the two while-loops that keep a set inside its catalogue bounds. A watch
 * that built "roughly the same" session would give a different workout from the
 * phone for the same account on the same day, and the difference would be
 * invisible until somebody counted.
 *
 * The phone remains the place to CHANGE any of this; this is a mirror, and the
 * comments below say what each piece is for so the mirror can be re-checked.
 */
object SessionBuilder {

    /** Levels run 1..5, so the top of the range is 5. */
    private const val TOP_LEVEL = 5

    /**
     * How long one exercise runs at a given level.
     *
     * Interpolates between the exercise's own min and max by level, then rounds
     * to a WHOLE number of its cycles - a half-finished squeeze is not a thing
     * anybody can do - while keeping the result inside the catalogue's bounds
     * at both ends.
     */
    fun durationForLevel(slug: String, levelNumber: Int): Double {
        val cycle = Catalogue.cycleSeconds(slug)
        if (cycle <= 0.0) return 30.0

        val (minSec, maxSec) = Catalogue.durationBounds(slug)
        val pct = if (TOP_LEVEL > 1) {
            max(0.0, min(1.0, (levelNumber - 1).toDouble() / (TOP_LEVEL - 1)))
        } else 0.0
        val target = minSec + (maxSec - minSec) * pct

        var cycles = max(1, (target / cycle).roundToInt())
        // Never blow past the maximum on rounding...
        while (cycles > 1 && cycles * cycle > maxSec + 0.01) cycles--
        // ...and symmetrically, never fall short of the minimum. Only the
        // maximum used to be guarded on the phone, which let a rounded set come
        // out shorter than the catalogue said the exercise was.
        while (cycles * cycle < minSec - 0.01 && (cycles + 1) * cycle <= maxSec + 0.01) cycles++

        return ((cycles * cycle) * 10).roundToInt() / 10.0
    }

    private fun steps(slug: String, duration: Double): List<Segment> {
        val def = Catalogue.exercises[slug] ?: return emptyList()
        val cycle = Catalogue.cycleSeconds(slug)
        val reps = if (cycle > 0) max(1, ((duration / cycle) + 1e-6).toInt()) else 0
        val out = mutableListOf<Segment>()
        repeat(reps) { out += def.pattern }
        return out
    }

    /**
     * The day's session.
     *
     * @param completedDays how far through the plan this account is - it is
     *        what unlocks exercises.
     * @param entitled a free account is bounded twice: its day count stops
     *        early AND the pool is narrowed here, so a day count that arrived
     *        from a period when the account was paying still cannot produce a
     *        paid exercise.
     */
    fun buildDaily(completedDays: Int, levelNumber: Int, entitled: Boolean): Playlist {
        val level = Catalogue.level(levelNumber)
        val total = level.totalSessionSeconds.toDouble()
        val rest = level.restSeconds.toDouble()

        val unlocked = Catalogue.availableSlugs(completedDays, entitled)
            .mapNotNull { Catalogue.exercises[it] }
        if (unlocked.isEmpty()) return Playlist(emptyList(), emptyList(), 0.0)

        val pool = unlocked.shuffled()
        val sequence = mutableListOf<ExerciseDef>()
        val durations = mutableListOf<Double>()
        var acc = 0.0
        var idx = 0

        // 30 is the phone's own ceiling: a guard against a pool of very short
        // exercises filling a long level with an unreadable number of blocks.
        while (sequence.size < 30) {
            var candidate = pool[idx % pool.size]
            // Never the same exercise twice in a row while there is an
            // alternative - back-to-back identical blocks read as a stutter.
            if (sequence.isNotEmpty() && sequence.last().slug == candidate.slug && pool.size > 1) {
                idx++
                candidate = pool[idx % pool.size]
            }

            val duration = durationForLevel(candidate.slug, levelNumber)
            val addition = (if (sequence.isEmpty()) 0.0 else rest) + duration

            // Stop at whichever side of the target is CLOSER, rather than
            // always stopping short: a session two seconds over is nearer the
            // level's length than one twenty seconds under.
            if (sequence.isNotEmpty() && acc + addition > total) {
                val over = (acc + addition) - total
                val under = total - acc
                if (over >= under) break
            }

            durations += duration
            sequence += candidate
            acc += addition
            idx++

            if (acc >= total) break
        }

        val out = mutableListOf<PlayStep>()
        val exercises = mutableListOf<String>()
        var finalAcc = 0.0

        sequence.forEachIndexed { i, exercise ->
            if (i > 0 && rest > 0) {
                out += PlayStep(
                    phase = "relax",
                    label = Catalogue.label("catalogue.steps.rest"),
                    seconds = rest,
                    from = 0.0,
                    to = 0.0,
                    slug = "rest",
                )
                finalAcc += rest
            }

            steps(exercise.slug, durations[i]).forEach { s ->
                out += PlayStep(
                    phase = s.phase,
                    label = Catalogue.label(s.labelKey),
                    seconds = s.seconds,
                    from = s.from,
                    to = s.to,
                    slug = exercise.slug,
                )
                finalAcc += s.seconds
            }

            if (exercise.slug !in exercises) exercises += exercise.slug
        }

        return Playlist(out, exercises, (finalAcc * 10).roundToInt() / 10.0)
    }

    /**
     * Every exercise at every level, as this app computes it.
     *
     * Exists so "are the exercises the same as the phone?" can be answered by
     * DIFFING rather than by looking. It emits the real Kotlin output - the same
     * `durationForLevel` and `steps` a session uses - which is the only version
     * worth checking; a transcription of this logic into another language would
     * only prove the transcription right.
     *
     * The phone dumps the same shape from its own test runner, so the two files
     * can be compared byte for byte. See scripts/compare-exercises.cjs.
     *
     * Debug builds only: `MainActivity` guards the call behind BuildConfig.DEBUG,
     * so nothing here reaches a release.
     */
    /**
     * Which exercises are available at a given day count, on both tiers.
     *
     * Added because "Front Clamp appeared and it was not unlocked" is a
     * question about the GATE, and the step dump does not cover the gate at
     * all - it compares what an exercise does, not whether it should have been
     * offered. This compares the pool itself.
     */
    fun dumpGateForComparison(): String {
        val q = '"'
        val days = listOf(0, 1, 2, 3, 5, 7, 14, 20, 36, 50, 109, 200)
        val out = StringBuilder()
        out.appendLine("{")
        days.forEachIndexed { di, d ->
            out.appendLine("  $q$d$q: {")
            listOf(true, false).forEachIndexed { ei, entitled ->
                val slugs = Catalogue.availableSlugs(d, entitled)
                val cells = slugs.joinToString(", ") { "$q$it$q" }
                val comma = if (ei == 0) "," else ""
                out.appendLine("    $q$entitled$q: [$cells]$comma")
            }
            out.appendLine(if (di < days.size - 1) "  }," else "  }")
        }
        out.appendLine("}")
        return out.toString()
    }

    fun dumpAllForComparison(): String {
        val q = '"'
        val out = StringBuilder()
        out.appendLine("{")
        val slugs = Catalogue.exercises.values.sortedBy { it.sortOrder }.map { it.slug }
        slugs.forEachIndexed { si, slug ->
            out.appendLine("  $q$slug$q: {")
            for (level in 1..5) {
                val duration = durationForLevel(slug, level)
                val steps = steps(slug, duration)
                val cells = steps.joinToString(", ") { st ->
                    "$q${st.phase} ${trim(st.seconds)} ${trim(st.from)}->${trim(st.to)}$q"
                }
                val comma = if (level < 5) "," else ""
                out.appendLine("    $q$level$q: {${q}duration$q: ${trim(duration)}, ${q}steps$q: [$cells]}$comma")
            }
            out.appendLine(if (si < slugs.size - 1) "  }," else "  }")
        }
        out.appendLine("}")
        return out.toString()
    }

    /** Numbers as the phone's JSON writes them: 29 not 29.0, 2.4 stays 2.4. */
    private fun trim(v: Double): String =
        if (v == v.toLong().toDouble()) v.toLong().toString() else v.toString()
}
